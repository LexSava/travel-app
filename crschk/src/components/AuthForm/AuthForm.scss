@import "../../variables.scss";

.auth {
  &-modal {
    background-color: rgba($color: #000000, $alpha: 0.3);
    position: absolute;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    z-index: 9;
  }

  &-form {
    position: relative;
    height: 100%;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: center;

    &--title {
      text-align: center;
      text-transform: uppercase;
      margin-bottom: 15px;
    }

    &--close-button {
      position: absolute;
      right: 2%;
      top: 2%;
      height: 36px;
      padding: 0 10px;
      border: none;
      border-radius: 18px;
      background-color: $color-focus;
      color: $color-text-light;
      text-transform: uppercase;
    }

    &--input {
      width: 100%;
      height: 36px;
      margin-bottom: 5px;

      background-color: $color-bg-input;

      border-radius: 18px;
      border: none;

      font-size: 18px;
      padding: 0 30px 0 15px;

      &:focus {
        outline: none;
        background-color: rgb(246, 246, 246);
      }

      &_file {
        cursor: pointer;
        width: 150px;
        height: 150px;
        margin-top: 20px;

        display: flex;
        flex-direction: column;
        justify-content: center;
        align-items: center;

        background-color: $color-bg-input;
        border-radius: 50%;
        text-transform: uppercase;
        color: $color-focus;
        font-weight: 900;

        background-size: cover;
      }
    }

    &-container {
      width: auto;
      height: auto;
      background-color: $color-body;
      padding: 30px;
      border-radius: 0.5rem;
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);

      .close {
        position: absolute;
        right: 20px;
        top: 20px;
        font-size: 24px;
        font-weight: 900;
        cursor: pointer;

        &:hover {
          color: $color-focus;
        }
      }
    }

    &_submit {
      height: 36px;
      padding: 0 10px;
      border: none;
      border-radius: 18px;
      background-color: $color-focus;
      color: $color-text-light;
      text-transform: uppercase;
      margin-top: 20px;

      &:hover {
        filter: brightness(1.2);
        cursor: pointer;
      }
    }
  }
}

input[type="file"] {
  display: none;
}
