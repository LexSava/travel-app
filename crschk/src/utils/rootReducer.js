import actions from "./actions";

const getData = (key, initialValue) => {
  const data = localStorage.getItem(key);

  if (!data) {
    localStorage.setItem(key, initialValue);
  }

  let result = data;
  try {
    result = JSON.parse(data);
  } catch {
    result = data;
  }

  return result ? result : initialValue;
};

export const initialState = {
  locale: getData("lang", "ru_RU"),
  dict: {},
  user: getData("user", null),
  searchValue: "",
  authForm: {
    isFormOpen: false,
    isSignup: true,
  },
};

export default function rootReducer(state = initialState, action) {
  switch (action.type) {
    case actions.SET_LOCALE: {
      localStorage.setItem("lang", action.payload);
      return {
        ...state,
        locale: action.payload,
      };
    }
    case actions.ADD_LOCALE: {
      return {
        ...state,
        dict: action.payload,
      };
    }
    case actions.SET_USER: {
      localStorage.setItem("user", JSON.stringify(action.user));
      return {
        ...state,
        user: getData("user", null),
      };
    }
    case actions.REMOVE_USER: {
      localStorage.setItem("user", JSON.stringify(null));
      return {
        ...state,
        user: getData("user", null),
      };
    }
    case actions.SEARCH: {
      return {
        ...state,
        searchValue: action.payload,
      };
    }
    case actions.SET_AUTHFORM: {
      return {
        ...state,
        authForm: {
          ...state.authForm,
          ...action.payload,
        },
      };
    }
    default:
      return state;
  }
}
