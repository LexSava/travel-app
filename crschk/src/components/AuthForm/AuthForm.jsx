import React, { useRef, useState } from "react";
import { useSelector } from "react-redux";
import { url, signin, signup } from "../../utils/api";

import "./AuthForm.scss";

const AuthForm = ({ isSignup, setUser, closeForm }) => {
  const { dict } = useSelector((state) => state);

  const [imgUrl, setImgUrl] = useState(
    "https://www.pinclipart.com/picdir/big/15-154296_gender-neutral-user-account-icon-png-clipart.png"
  );
  const [loginValue, setLoginValue] = useState("");
  const [loginPlaceHolder, setLoginPlaceHolder] = useState(dict.NAME);

  const reader = new FileReader();
  reader.onload = (e) => {
    setImgUrl(e.target.result);
  };

  const submit = async (e) => {
    e.preventDefault();
    const data = new FormData(e.target);
    const res = isSignup ? await signup(data) : await signin(data);
    const resStatus = await res.status;
    const resUser = await res.user;
    if (resStatus === 403) {
      setLoginValue("");
      isSignup
        ? setLoginPlaceHolder(dict.NAME_TAKEN)
        : setLoginPlaceHolder(dict.NAME_INCORRECT);
    } else if (resStatus === 200) {
      if (resUser) {
        setUser({
          login: loginValue,
          avatar: `${resUser.avatar}`,
        });
      } else {
        setUser({
          login: loginValue,
          avatar: `${url}statics/users/${loginValue.toLowerCase()}${imgUrl.substr(
            imgUrl.lastIndexOf(".")
          )}`,
        });
      }
      closeForm();
    }
  };

  return (
    <div className="auth-modal" onClick={closeForm}>
      <div className="auth-form-container">
        <div className="close" onClick={closeForm}>X</div>
        <h2 className="auth-form--title">{isSignup ? dict.SIGNUP : dict.LOGIN}</h2>
        <form className="auth-form" onSubmit={submit}>
          <input
            name="login"
            type="text"
            required
            placeholder={loginPlaceHolder}
            value={loginValue}
            className="auth-form--input"
            onChange={(e) => {
              setLoginValue(e.target.value);
            }}
          />
          <input type="password" name="password" className="auth-form--input" required placeholder={dict.PASSWORD} />
          {isSignup && (
            <label
              className="add-avatar-btn auth-form--input_file"
              style={{ backgroundImage: `url(${imgUrl})` }}
            >
              <input
                type="file"
                accept="image/*"
                name="avatar"
                onChange={(e) => {
                  reader.readAsDataURL(e.target.files[0]);
                }}
              />
              {dict.ADD_AVATAR}
            </label>
          )}
          <input type="submit" className="auth-form_submit" value={dict.SUBMIT} />
        </form>
      </div>
    </div>
  );
};

export default AuthForm;
