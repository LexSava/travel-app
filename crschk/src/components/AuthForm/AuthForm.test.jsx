import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";

import rootReducer from "../../utils/rootReducer";
import "babel-polyfill";

import AuthForm from "./AuthForm";
import { createStore } from "redux";

describe("Auth form", () => {
  it("Should render signup form", () => {
    const tree = renderer
      .create(
        <Provider store={createStore(rootReducer)}>
          <AuthForm isSignup />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should render signin form", () => {
    const tree = renderer
      .create(
        <Provider store={createStore(rootReducer)}>
          <AuthForm />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
