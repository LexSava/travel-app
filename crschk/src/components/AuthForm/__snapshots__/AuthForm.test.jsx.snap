// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth form Should render signin form 1`] = `
<div
  className="auth-modal"
>
  <div
    className="auth-form-container"
  >
    <div
      className="close"
    >
      X
    </div>
    <h2
      className="auth-form--title"
    />
    <form
      className="auth-form"
      onSubmit={[Function]}
    >
      <input
        className="auth-form--input"
        name="login"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
      <input
        className="auth-form--input"
        name="password"
        required={true}
        type="password"
      />
      <input
        className="auth-form_submit"
        type="submit"
      />
    </form>
  </div>
</div>
`;

exports[`Auth form Should render signup form 1`] = `
<div
  className="auth-modal"
>
  <div
    className="auth-form-container"
  >
    <div
      className="close"
    >
      X
    </div>
    <h2
      className="auth-form--title"
    />
    <form
      className="auth-form"
      onSubmit={[Function]}
    >
      <input
        className="auth-form--input"
        name="login"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
      <input
        className="auth-form--input"
        name="password"
        required={true}
        type="password"
      />
      <label
        className="add-avatar-btn auth-form--input_file"
        style={
          Object {
            "backgroundImage": "url(https://www.pinclipart.com/picdir/big/15-154296_gender-neutral-user-account-icon-png-clipart.png)",
          }
        }
      >
        <input
          accept="image/*"
          name="avatar"
          onChange={[Function]}
          type="file"
        />
      </label>
      <input
        className="auth-form_submit"
        type="submit"
      />
    </form>
  </div>
</div>
`;
